### DISCUSSION TITLE
Dynamic validator for custom type

### DISCUSSION THREAD

**Discussion by Credics** [pydantic/pydantic#2993] [Question] [answered]
I create a model with `create_model`. For this process, I need a validator for a custom type. Now here is the problem:

The validator is dynamic, so I cannot simply add a `__get_validators__` method to my type. Also, the type can occur in a complex schema (as an argument of a generic type or in a nested model). Therefore, I cannot use the `__validators__` parameter of `create_model` or use this #2873 approach (as far as I understood they only work for fields).

I came up with two ideas, which are both somewhat 'hacky':

1. dynamically add the `__get_validators__` method to my type and remove it afterwards.
2. dynamically add the type and the validator to `pydantic.validators._VALIDATORS` and remove it afterwards.

Does anybody have a cleaner solution?

*Edit*: Should have added an example.

A minimal example would look something like this
```python
class CustomType:
    @classmethod
    def validate(cls, obj, manager):
        if not isinstance(obj, cls):
            raise TypeError
        manager.validate(obj)
        return obj

fields = {
    x1: CustomType,
    x2: List[Tuple[int, CustomType]],
}
model = create_model('my_model', **fields)
```
where the `manager` is created shortly before `create_model` is called. While I could use `functools.partial` to provide the `manager`, I do not know where to put that wrapped validator.

**Comment by lesnik512** [comment]
Hi, examples would be great.
And why exactly can't you simply add a __get_validators__ method to your type?
What does it mean that validator is dynamic?

**Reply by Credics** [reply]
I hope the edit makes it clear.

**Comment by lesnik512** [comment, marked as answer]
what if you declare `manager` above `CustomType`? Sth like that

```python
manager = None

class CustomType:
    @classmethod
    def validate(cls, obj):
        if not isinstance(obj, cls):
            raise TypeError
        manager.validate(obj)
        return obj

fields = {
    x1: CustomType,
    x2: List[Tuple[int, CustomType]],
}
manager = // some assignment
model = create_model('my_model', **fields)
```

**Reply by Credics** [reply]
That should work. Thanks :) Any idea how to avoid using a global variable?

**Reply by lesnik512** [reply]
You can add class property, I think. I'm not sure if that's better, but I have no other ideas

```python
class CustomType:
    manager = None
    @classmethod
    def validate(cls, obj):
        if not isinstance(obj, cls):
            raise TypeError
        cls.manager.validate(obj)
        return obj

fields = {
    x1: CustomType,
    x2: List[Tuple[int, CustomType]],
}
CustomType.manager = // some assignment
model = create_model('my_model', **fields)
```

**Reply by Credics** [reply]
I'll try that. Thanks again.

**Reply by lesnik512** [reply]
Glad to help) If problem is solved, I would be grateful if you mark answer as correct
